src/battle_script_commands.c

```diff
@@ -3799,6 +3799,32 @@ void SetMoveEffect(bool32 primary, u32 certain)
                     gBattlescriptCurrInstr = BattleScript_SpikesActivates;
                 }
                 break;
+            case MOVE_EFFECT_TRIPLE_ARROWS:
```

Comment:
Defense drop and flinch can both happen. It's not one or the other.

Reply:
This is ugly, but it works, but it's ugly. What I did in this PR also feels ugly tbh, but this feels even uglier.
Maybe it's just me, what do you think? 😂 
```c
case MOVE_EFFECT_TRIPLE_ARROWS:
                {
                    u8 randomChance = RandomUniform(RNG_TRIPLE_ARROWS, 1, 10);
                    bool8 shouldLowerDefense = FALSE;
                    bool8 shouldFlinch = TRUE;

                    if (randomChance <= 5) // Chance to reduce a foe's Defense by 1 stat stage.
                        shouldLowerDefense = TRUE;
                    if (randomChance > 5 && randomChance <= 8) // Chance to cause a foe to flinch.
                        shouldFlinch = TRUE;

                    if (shouldLowerDefense && shouldFlinch)
                    {
                        BattleScriptPush(gBattlescriptCurrInstr + 1);
                        gBattlescriptCurrInstr = BattleScript_ReduceDefenseAndFlinch;
                    }
                    else
                    {
                        if (shouldLowerDefense)
                        {
                            BattleScriptPush(gBattlescriptCurrInstr + 1);
                            gBattlescriptCurrInstr = BattleScript_DefDown;
                        }
                        else if (shouldFlinch)
                        {
                            if (battlerAbility == ABILITY_INNER_FOCUS && (primary == TRUE || certain == MOVE_EFFECT_CERTAIN))
                            {
                                gLastUsedAbility = ABILITY_INNER_FOCUS;
                                gBattlerAbility = gEffectBattler;
                                RecordAbilityBattle(gEffectBattler, ABILITY_INNER_FOCUS);
                                gBattlescriptCurrInstr = BattleScript_FlinchPrevention;
                            }
                            else
                            {
                                if (GetBattlerTurnOrderNum(gEffectBattler) > gCurrentTurnActionNumber)
                                    gBattleMons[gEffectBattler].status2 |= sStatusFlagsForMoveEffects[MOVE_EFFECT_FLINCH];
                                gBattlescriptCurrInstr++;
                            }
                        }
                    }
                }
                break;
            }
```
```
BattleScript_ReduceDefenseAndFlinch::
	modifybattlerstatstage BS_TARGET, STAT_DEF, DECREASE, 1, BattleScript_DefDown_Ret, ANIM_ON
	setmoveeffect MOVE_EFFECT_FLINCH
	seteffectprimary
	goto BattleScript_MoveEnd
```

Reply:
```
            if (randomLowerDefenseChance)
                shouldLowerDefense = TRUE;
            if (randomFLinchChance)
                shouldFlinch = TRUE;
```
No need to make another variable. Just use `randomLowerDefenseChance` and `randomFLinchChance`.

---

Instead of making a combined battle script, just push them both to the battle script stack. When one finishes it will pop back down to the next effect.

Reply:
> No need to make another variable. Just use `randomLowerDefenseChance` and `randomFLinchChance`.

In the same way I did it before, you mean?

```c
            u8 randomChance = RandomUniform(RNG_TRIPLE_ARROWS, 1, 100);
            bool8 shouldLowerDefense = FALSE, shouldFlinch = FALSE;
    
            if (randomChance <= 50)
                shouldLowerDefense = TRUE;
            if (randomChance >= 51 && randomChance <= 80)
                shouldFlinch = TRUE;
```

Reply:
Sorry, I've missed this whole randomness chat because I've been away from the PC doing lots of house move tasks. I think maybe you'd be better off with something like:
```c
bool8 shouldLowerDefense = RandomPercentage(RNG_TRIPLE_ARROWS_DEFENSE, 30);
bool8 shouldFlinch = RandomPercentage(RNG_TRIPLE_ARROWS_FLINCH, 50);
```
The reason for this is that when any `PASSES_RANDOMLY` tests run for something like `RandomUniform(tag, 1, 100)` the tests will run 100 times, once for each value 1..100. Whereas `RandomPercentage(tag, t)` will run just two times, returning a `FALSE` and a `TRUE`.

I'd also say `RandomPercentage` is a bit more straightforward for newcomers to wrap their heads around than using `RandomUniform` to build a percentage, but mostly my concern is the first thing :)

I've also split the tag into two here, just in case somebody wants to test with defense drop but not flinch, or vice-versa. That's honestly quite minor too, but generally speaking if two `RandomX` calls lead to different effects, I'd expect them to have different tags.

Reply:
> Sorry, I've missed this whole randomness chat because I've been away from the PC doing lots of house move tasks. I think maybe you'd be better off with something like:
> 
> ```c
> bool8 shouldLowerDefense = RandomPercentage(RNG_TRIPLE_ARROWS_DEFENSE, 50);
> bool8 shouldFlinch = RandomPercentage(RNG_TRIPLE_ARROWS_FLINCH, 30);
> ```

Skeli did suggest to use a single variable for the randomChance though, basically like this PR currently has :eyes:

Reply:
I wouldn't do it with a single random chance variable because the two effects are independent, but if you wanted to do that I'd recommend `RandomWeighted`—the four possibilities are:

No Flinch, No Defense: 70%*50% = 35%
Flinch, No Defense: 30%*50% = 15%
No Flinch, Defense: 70%*50% = 35%
Flinch, Defense: 30%*50% = 15%

Which translates to something like:
```c
enum TripleArrowsEffect { TRIPLE_ARROWS_NONE, TRIPLE_ARROWS_FLINCH, TRIPLE_ARROWS_DEFENSE_DOWN, TRIPLE_ARROWS_FLINCH_DEFENSE_DOWN };

enum TripleArrowsEffect effect = RandomWeighted(RNG_TRIPLE_ARROWS, 35, 15, 35, 15);
```
This is better than doing `RandomUniform(RNG_TRIPLE_ARROWS, 1, 100)` and checking for 1-35, 36-50, 51-85, and 86-100 because of the `PASSES_RANDOMLY` test performance thing I mentioned before (`RandomUniform` would run 100 battles, whereas the above would run 4).

For contrast, if the effects were independent I would use `RandomWeighted` because then you can have something like:
```c
enum TripleArrowsEffect { TRIPLE_ARROWS_NONE, TRIPLE_ARROWS_FLINCH, TRIPLE_ARROWS_DEFENSE_DOWN };

enum TripleArrowsEffect effect = RandomWeighted(RNG_TRIPLE_ARROWS, 20, 30, 50);
```
Which I think is quite nice for saying that the percentages are 20, 30, and 50.

(Note you don't have to use an `enum` here, you could just use the numbers 0-3 or 0-2 which is actually what `RandomWeighted` would return)

Reply:
> I wouldn't do it with a single random chance variable because the two effects are independent

What I'm doing right now outside of this PR, as I showed above 1h ago, is to have `bool`s that confirm if either effect should activate, so having a single variable to handle the random chance shouldn't be a problem per se.

Now, how does `RandomPercentage` work exactly? You say that it returns `TRUE` or `FALSE` and it only reads a single digit to determine the max threshold for a percentage, yeah?

Does that mean that something like `RandomPercentage(RNG_TRIPLE_ARROWS_DEF_DOWN, 50)` would return `TRUE` if a number from 0 to 50 is generated, and `FALSE` if a number from 51 to 100 is generated instead?

Because that would mean I'd need to use 2 variables for random chances, just like I did 14hs ago.

*Which I personally don't really have a problem with, but is the entire point of the discussion.*

Reply:
> Now, how does RandomPercentage work exactly? You say that it returns TRUE or FALSE and it only reads a single digit to determine the max threshold for a percentage, yeah?
>
> Does that mean that something like `RandomPercentage(RNG_TRIPLE_ARROWS_DEF_DOWN, 50)` would return TRUE if a number from 0 to 50 is generated, and FALSE if a number from 51 to 100 is generated instead?
Yeah, so for example `RandomPercentage(tag, 30);` returns `TRUE` 30% of the time, and `FALSE` the other 70% of the time. In your example it's 50-50. You can think of it as being similar to generating a number from 1-100 (or 0-99 or whatever) and comparing it, although technically it works differently internally and that's why it has better performance when used with `PASSES_RANDOMLY` in the tests :)

And yeah, I think you should use two variables for the random chances :)
But if for some reason people think you should just use one variable, the first `RandomWeighted` example is the best way to achieve that, because it has good test performance. I just don't think the code is as easy to modify when it's written like that (and I guess you'd agree because you didn't do it that way to begin with either!)

Reply:
> Skeli did suggest to use a single variable for the randomChance though
I meant don't make unnecessary extra variables, just use the variables you already have.
Eg. simplify to something like:
```
u8 randomLowerDefenseChance = RandomUniform(RNG_TRIPLE_ARROWS, 1, 50);
u8 randomFLinchChance = RandomUniform(RNG_TRIPLE_ARROWS, 1, 30);

gBattlescriptCurrInstr++;

if (randomLowerDefenseChance)
{
    BattleScriptPushCursor();
    gBattlescriptCurrInstr = BattleScript_DefDown;
}

if (randomFLinchChance
&& battlerAbility != ABILITY_INNER_FOCUS
&& GetBattlerTurnOrderNum(gEffectBattler) > gCurrentTurnActionNumber)
{
    gBattleMons[gEffectBattler].status2 |= sStatusFlagsForMoveEffects[MOVE_EFFECT_FLINCH];
}
```

What mgriffin said about the function you were using for randomness is not related to what I wrote.

Reply:
> > Skeli did suggest to use a single variable for the randomChance though
> 
> I meant don't make unnecessary extra variables, just use the variables you already have. Eg. simplify to something like:
> 
> ```
> u8 randomLowerDefenseChance = RandomUniform(RNG_TRIPLE_ARROWS, 1, 50);
> u8 randomFLinchChance = RandomUniform(RNG_TRIPLE_ARROWS, 1, 30);
> 
> gBattlescriptCurrInstr++;
> 
> if (randomLowerDefenseChance)
> {
>     BattleScriptPushCursor();
>     gBattlescriptCurrInstr = BattleScript_DefDown;
> }
> 
> if (randomFLinchChance
> && battlerAbility != ABILITY_INNER_FOCUS
> && GetBattlerTurnOrderNum(gEffectBattler) > gCurrentTurnActionNumber)
> {
>     gBattleMons[gEffectBattler].status2 |= sStatusFlagsForMoveEffects[MOVE_EFFECT_FLINCH];
> }
> ```
> 
> What mgriffin said about the function you were using for randomness is not related to what I wrote. I can't comment on what he said as I don't use your system.

Sorry, I misread completely.

Uh, I can't do it exactly like you suggested because that doesn't work, but I'll keep trying and see what I can manage.

The Inner Focus bit and its battle script call needs to stay though.

Reply:
> The Inner Focus bit and its battle script call needs to stay though.

It's not necessary. This is a purely secondary effect so the case `(primary == TRUE || certain == MOVE_EFFECT_CERTAIN)` will never evaluate to true. You could keep it in if you _really_ want, and in that case you'd just need to push it to the stack as an extra battle script to play (but really, it's unnecessary).

Reply:
It may not be necessary, but is removing it something that is accurate to the official games?

I would imagine that if the flinching happened, the code in SV would try to activate Inner Focus, and if that happens, the Player needs to be informed about it.

I'll try to test it out myself.

Reply:
I couldn't get things working, so I sticked to what I showed above while also adding a check for Serene Grace and an Inner Focus check in the Def Down + Flinch battle script.